Add support for permanent locked vesting accounts

clevinson opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: #8283

While the SDK currently supports 3 different types of vesting accounts (delayed, continuous, periodic), it only exposes one Msg for creating vesting accounts, parameterized by a `delayed` boolean, which either creates a "delayed" or "continuous" vesting account.

After speaking with @aaronc about this today, I've realized that we don't actually need `sdk.Msg` support for creation of `PermanentLockedVestingAccount`. We discussed a bit the issues with the current vesting logic (as described in #8528), and as a result, i'd like to proceed with the most minimal solution in this PR.

This simply adds support for a new `PermanentLockedVestingAccount` type. There shouldn't really be any need for creating these accounts after genesis (as @aaronc noted below). In Regen's own use case, we will setup these accounts to hold locked funds of a Group account (for our Community Staking DAOs), and even with these DAOs not fully realized at genesis, we can instantiate the relevant group accounts and correpsonding `PermanentLockedVestingAccount`s at genesis.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

aaronc commented:
Not sure we even need to create this sort of thing after genesis, but maybe... Also I'd like to see a more generic way to deal with vesting eventually but that's out of scope.

clevinson commented:
@aaronc As it currently stands (on the Regen side atleast), we will not have our Community Staking DAOs setup prior to genesis. It's been my understanding that Regen Foundation would hold all tokens for community staking dao's, and distribute them as the organizations are brought onboard (after mainnet).

clevinson commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -570,3 +570,69 @@ func (dva DelayedVestingAccount) String() string {
+// NewPermanentLockedVestingAccountRaw creates a new PermanentLockedVestingAccount object from BaseVestingAccount
+func NewPermanentLockedVestingAccountRaw(bva *BaseVestingAccount) *PermanentLockedVestingAccount {
+	// ensure EndTime is set to -1, as PermanentLockedVestingAccount's do not have an EndTime
+	bva.EndTime = -1
```
Seems kind of hacky... but otherwise we may unwillingly store an `EndTime` value from an inherited BaseVestingAccount. This `EndTime` doesn't actually do anything... as all the logic is handled in the implementations of the `PermanentLockedVestingAccount`'s functions.

aaronc replied:
Why do we need to embed BaseVestingAccount at all? If we don't embed we won't have this problem

clevinson replied:
Even if we drop the embedded `BaseVestingAccount` we will need to implement `GetStartTime()` and `GetEndTime()` somehow to implement the exported VestingAccount interface [see here](https://github.com/cosmos/cosmos-sdk/blob/master/x/auth/vesting/exported/exported.go#L35-L36).

@aaronc were you suggesting we refactor that interface as well? Or a strategy where we actually don't consider `PermanentLockedVestingAccount` a vesting account at all? I guess there isn't really anything "vesting" about it. So maybe decoupling it entirely from the vesting account interface is prefereble... I'd have to investigate what that would look like.

aaronc replied:
Let's keep it simple for now IMHO even if we have to do something weird for the interface. But we should check consequences on returning invalidate times. When we get around to it I want to refactor vesting accounts into something much more modular.

amaurym replied:
> Can it embed BaseAccount, + another OriginalVesting field?

My suggestion above would require duplicating a lot of the `BaseVestingAccount` logic into `PermanentLockedVestingAccount`, so maybe not the best idea.

I think @clevinson what you're doing here seems like the most straightforward way right now. Internally in the SDK, we don't use `GetEndTime` ever (except in the local ValidateBasic). If a user queries this account (via gRPC or whatever), then returning `-1` on that field should be pretty clear that it doesn't have an EndTime.

clevinson commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -578,6 +578,152 @@ func TestTrackUndelegationPeriodicVestingAcc(t *testing.T) {
+	// receive some coins
+	// require that only received coins are spendable since the original coins
+	// are all locked
+	lockedCoins = plva.LockedCoins(endTime)
```
This seems like an inaccurate assertion. There aren't any "coins received" in the test case. Same goes for all the `TestSpendableCoins{SomeOther}VestingAcc()` implementations though. Let me know if i'm missing something here, or if we should actually test for receiving some coins. This would require setting up a bank keeper iiuc?

amaurym replied:
I had a look at the other Spendable tests, the "receive some coins" part always seems to just be a copy/pate of the previous line... I'm a bit confused here too, maybe @alexanderbez has an idea?

alexanderbez replied:
So these comments are remnants from ADR-004 implementation with the test modifications [here](https://github.com/cosmos/cosmos-sdk/pull/5572/files#diff-79cdc833ea7de9c9dcb705ff51d3b0d2648ce639a7438c85e5a8cea2671577b1). Looks like I forgot to remove the `// receive some coins` comments :-/

Apologies.

alexanderbez replied:
Before ADR-004, these tests did have accounts receive coins, that's why those comments existed. They should be removed now 👍

clevinson commented:
@aaronc I've updated this to remove the `Msg` support. Let's just stick with setting these up with Group accounts for now. I've updated the description in this PR accordingly.

clevinson commented on proto/cosmos/vesting/v1beta1/tx.proto:
```diff
@@ -27,5 +27,21 @@ message MsgCreateVestingAccount {
+// MsgCreatePermanentLockedAccount defines a message that creates a permanent locked
+// account. The original coin balance of this account can be used for voting and delegation, but can
+// never be traded or sent to another account.
+message MsgCreatePermanentLockedAccount {
```
Nono - I had added in a previous Pr. Will drop this. Thx for the catch !

aaronc commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -566,3 +566,69 @@ func (dva DelayedVestingAccount) String() string {
+}
+
+// NewPermanentLockedVestingAccount returns a PermanentLockedVestingAccount
+func NewPermanentLockedVestingAccount(baseAcc *authtypes.BaseAccount, originalVesting sdk.Coins) *PermanentLockedVestingAccount {
```
```suggestion
func NewPermanentLockedVestingAccount(baseAcc *authtypes.BaseAccount, coins sdk.Coins) *PermanentLockedVestingAccount {
```

aaronc commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -566,3 +566,69 @@ func (dva DelayedVestingAccount) String() string {
+
+// LockedCoins returns the set of coins that are not spendable (i.e. locked).
+func (plva PermanentLockedVestingAccount) LockedCoins(blockTime time.Time) sdk.Coins {
+	return plva.BaseVestingAccount.LockedCoinsFromVesting(plva.GetVestingCoins(blockTime))
```
```suggestion
	return nil
```

amaurym replied:
Your suggestion is incorrect imo, `LockedCoins` means vesting coins that are not delegated.

aaronc commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -566,3 +566,69 @@ func (dva DelayedVestingAccount) String() string {
+// values for the amount of delegated vesting, delegated free, and reducing the
+// overall amount of base coins.
+func (plva *PermanentLockedVestingAccount) TrackDelegation(blockTime time.Time, balance, amount sdk.Coins) {
+	plva.BaseVestingAccount.TrackDelegation(balance, plva.GetVestingCoins(blockTime), amount)
```
```suggestion
	plva.BaseVestingAccount.TrackDelegation(balance, plva.OriginalVesting, amount)
```

aaronc commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -566,3 +566,69 @@ func (dva DelayedVestingAccount) String() string {
+}
+
+// GetStartTime returns zero since a permanent locked vesting account has no start time.
+func (plva PermanentLockedVestingAccount) GetStartTime() int64 {
```
To me it still makes sense to just override `GetEndTime` too. Relying on it being `-1` in the constructor feels a bit hacky.

amaurym replied:
Added

robert-zaremba replied:
I think it would make sense to use `GetEndTime() = 0` - it would match the `Validate` method.

amaurym replied:
OK, changed to 0. I also prefer this.

aaronc commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -89,10 +89,6 @@ func TestSpendableCoinsContVestingAcc(t *testing.T) {
-	// require that all vested coins (50%) are spendable plus any received
-	lockedCoins = cva.LockedCoins(now.Add(12 * time.Hour))
-	require.Equal(t, sdk.Coins{sdk.NewInt64Coin(feeDenom, 500), sdk.NewInt64Coin(stakeDenom, 50)}, lockedCoins)
 }
```
Why were these deleted?

amaurym replied:
They are copy-paste from 4 lines above.

aaronc commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -241,12 +237,6 @@ func TestSpendableCoinsDelVestingAcc(t *testing.T) {
 	lockedCoins = dva.LockedCoins(now.Add(12 * time.Hour))
 	require.True(t, lockedCoins.IsEqual(origCoins))
 
-	// receive some coins
-	// require that only received coins are spendable since the account is still
-	// vesting
-	lockedCoins = dva.LockedCoins(now.Add(12 * time.Hour))
-	require.True(t, lockedCoins.IsEqual(origCoins))
```
why?

aaronc commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -453,11 +443,6 @@ func TestSpendableCoinsPeriodicVestingAcc(t *testing.T) {
 	// require that all still vesting coins (50%) are locked
 	lockedCoins = pva.LockedCoins(now.Add(12 * time.Hour))
 	require.Equal(t, sdk.Coins{sdk.NewInt64Coin(feeDenom, 500), sdk.NewInt64Coin(stakeDenom, 50)}, lockedCoins)
-
-	// receive some coins
-	// require that all still vesting coins (50% of original) are locked plus any received
-	lockedCoins = pva.LockedCoins(now.Add(12 * time.Hour))
-	require.Equal(t, sdk.Coins{sdk.NewInt64Coin(feeDenom, 500), sdk.NewInt64Coin(stakeDenom, 50)}, lockedCoins)
```
why?

aaronc commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -578,6 +563,146 @@ func TestTrackUndelegationPeriodicVestingAcc(t *testing.T) {
+	origCoins := sdk.Coins{sdk.NewInt64Coin(feeDenom, 1000), sdk.NewInt64Coin(stakeDenom, 100)}
+	bacc := authtypes.NewBaseAccountWithAddress(addr)
+
+	// require no coins are vested until schedule maturation
```
comment makes no sense
```suggestion
	// require no coins are vested
```

aaronc commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -578,6 +563,146 @@ func TestTrackUndelegationPeriodicVestingAcc(t *testing.T) {
+	require.Nil(t, plva.DelegatedFree)
+	require.Nil(t, plva.DelegatedVesting)
+
+	// vest 50% and delegate to two validators
```
vest 50%?

amaurym commented on x/auth/vesting/types/common_test.go:
```diff
@@ -5,6 +5,5 @@ import (
 
 var (
-	app      = simapp.Setup(false)
-	appCodec = simapp.MakeTestEncodingConfig().Marshaler
```
unused var

aaronc replied:
Can we delete?

amaurym commented:
This PR is r4r again. Would appreciate if @aaronc and @alexanderbez could have a look 🙏

aaronc commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -523,6 +523,72 @@ func (dva DelayedVestingAccount) String() string {
+
+// LockedCoins returns the set of coins that are not spendable (i.e. locked).
+func (plva PermanentLockedVestingAccount) LockedCoins(_ time.Time) sdk.Coins {
+	return plva.BaseVestingAccount.LockedCoinsFromVesting(plva.OriginalVesting)
```
Why not just `return plva.OriginalVesting`?

amaurym replied:
My understanding is LockedCoins = OriginalVesting - DelegatedCoins, i.e. the coins that are unused.

aaronc replied:
Okay right

amaurym commented on proto/cosmos/vesting/v1beta1/vesting.proto:
```diff
@@ -71,3 +71,13 @@ message PeriodicVestingAccount {
+  option (gogoproto.goproto_getters)  = false;
+  option (gogoproto.goproto_stringer) = false;
+
+  BaseVestingAccount base_vesting_account = 1 [(gogoproto.embed) = true];
```
I changed `PermanentLockedVestingAccount` to `PermanentLockedAccount`, but I disagree with this one: `BaseVestingAccount` has a field `BaseAccount`. I dont want to do `permanentAccount.BaseAccount.BaseAccount.*`

amaurym commented:
In the latest commits, I changed:
- the semantics of `EndTime`: `0` means no end time, not `-1`
- rename PermanentLockedVestingAccount to PermanentLockedAccount

@robert-zaremba lmk if that lgty.
